### PR TITLE
bpo-46613: Add PyType_GetModuleByDef to the public & limited API

### PR DESCRIPTION
**Pull request by encukou**
- Remove the underscore from the name (everywhere)
- Add to limited API

https://bugs.python.org/issue46613

### CODE REVIEW

**Review by erlend-aasland — APPROVED** [review]

**Review by shihai1991 — APPROVED** [review]
Thanks, LGTM.

**Comment by shihai1991** [comment]
whatsnew/3.11.rst shoud be updated?

**Comment by erlend-aasland** [comment]
> whatsnew/3.11.rst shoud be updated?

Yes, good catch.

**Comment by encukou** [comment]
Here it is!
I ca't quite come up with a way to summarize it for people who haven't read up on the topic. Hopefully it's enough.

**Comment by vstinner** [comment]
Would it make sense to not add it to the limited API in Python 3.11, try to convert a few third party C extensions using it, and if everything goes well, add it to the limited API?

**Comment by encukou** [comment]
> Would it make sense to not add it to the limited API in Python 3.11, try to convert a few third party C extensions using it, and if everything goes well, add it to the limited API?

It would! OTOH, addition to the limited API is also easy to revert, and the function itself already proved useful in stdlib.
Should I remove the stable ABI addition for now?

**Comment by erlend-aasland** [comment]
I think that makes sense. AFAICS, only a handful of devs (Christian Heimes, Hai Shi, Petr, Victor, me) have been using this function in the CPython code base. In this case, maybe it would make sense to try it out in the wild first. Who could we ask?

File: Doc/c-api/type.rst [hunk 1]
```diff
@@ -166,6 +168,21 @@ Type Objects
 
    .. versionadded:: 3.9
 
+.. c:function:: void* PyType_GetModuleByDef(PyTypeObject *type, struct PyModuleDef *def)
```
Comment (vstinner) [inline Doc/c-api/type.rst]: The return type is `` PyObject*``, no?

```suggestion
.. c:function::  PyObject* PyType_GetModuleByDef(PyTypeObject *type, struct PyModuleDef *def)
```

File: Doc/c-api/type.rst [hunk 2]
```diff
@@ -166,6 +168,21 @@ Type Objects
...
+
+   Find the first superclass whose module was created from
+   the given PyModuleDef, and return that module.
+
+   If no module is found, sets :py:class:`TypeError` and returns ``NULL``.
```
Comment (vstinner) [inline Doc/c-api/type.rst]: I like "raises" to "sets" is correct, it's just a suggestion:

```suggestion
   If no module is found, raises a :py:class:`TypeError` and returns ``NULL``.
```

File: Doc/c-api/type.rst [hunk 3]
```diff
@@ -166,6 +168,21 @@ Type Objects
 
    .. versionadded:: 3.9
 
+.. c:function:: void* PyType_GetModuleByDef(PyTypeObject *type, struct PyModuleDef *def)
+
+   Find the first superclass whose module was created from
+   the given PyModuleDef, and return that module.
```
Comment (vstinner) [inline Doc/c-api/type.rst]:
```suggestion
   the given :ctype:`PyModuleDef` *def*, and return that module.
```

File: Doc/c-api/type.rst [hunk 4]
```diff
@@ -166,6 +168,21 @@ Type Objects
...
+
+   If no module is found, sets :py:class:`TypeError` and returns ``NULL``.
+
+   This function is intended to be used together with
+   :c:func:`PyModule_GetState()` to get module state from slot methods (such as
+   :c:member:`~PyTypeObject.tp_init` or :c:member:`~PyNumberMethods.nb_add`)
+   and other places where a method's defining class cannot be passed using the
+   :c:type:`PyCMethod` calling convention.
```
Comment (vstinner) [inline Doc/c-api/type.rst:182]: You may mention somewhere that this function only works on a type created by ``PyType_FromModuleAndSpec``.

File: Doc/c-api/type.rst (hunk 3 above)
Comment (encukou) [reply Doc/c-api/type.rst]: Sphinx puts a link in the signature, but why not repeat it.

File: Doc/c-api/type.rst (hunk 4 above)
Comment (encukou) [reply Doc/c-api/type.rst:182]: It works with any method that attaches the module to the class. Currently there's just one, but I've been adding new functions with extra arguments relatively fast. I'd rather not document one specific function.

**Review by vstinner — APPROVED** [review]
LGTM.

It's better with the correct return type in the doc :-)

I hope that we will be able to add it to the limited C API soon!

**Pull request merged**